File: LeetCode_783/main.cpp
/**
    LeetCode_783
    https://leetcode.com/problems/minimum-distance-between-bst-nodes/description/
    MatrixDeity, 2018.
*/

#include <iostream>
#include <functional>
#include <cmath>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int minDiffInBST(TreeNode* root) {
        const int MAX = 0x7FFFFFFF;
        int res = MAX;
        function<void(TreeNode*, int, int)> impl = [&impl, &res] (TreeNode* root, int mx, int mn) {
            if (root == nullptr)
                return;

            int cur = min(abs(mx - root->val), abs(root->val - mn));
            res = min(cur, res);
            impl(root->left, root->val, mn);
            impl(root->right, mx, root->val);
        };
        impl(root, MAX, MAX);
        return res;
    }
};

int main() {
    return 0;
}

File: LeetCode_783/main.go
/**
  LeetCode_783
  https://leetcode.com/problems/minimum-distance-between-bst-nodes/
  MatrixDeity, 2023.
*/

package leetcode783

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

const (
	big int = 1e6
)

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}

func minDiffInBST(root *TreeNode) int {
	res := big

	var impl func(node *TreeNode, mx int, mn int)
	impl = func(node *TreeNode, mx int, mn int) {
		if node == nil {
			return
		}

		minDiff := min(abs(mx-node.Val), abs(node.Val-mn))
		res = min(minDiff, res)
		impl(node.Left, node.Val, mn)
		impl(node.Right, mx, node.Val)
	}

	impl(root, -big, big)

	return res
}
